src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/tileentity/BannerDataBuilder.java

```diff
@@ -0,0 +1,90 @@
+package org.spongepowered.common.data.builder.manipulator.mutable.tileentity;
+
+import java.awt.Color;
+import java.util.List;
+
+import net.minecraft.item.EnumDyeColor;
+import net.minecraft.tileentity.TileEntityBanner;
+
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.DataManipulatorBuilder;
+import org.spongepowered.api.data.manipulator.immutable.tileentity.ImmutableBannerData;
+import org.spongepowered.api.data.manipulator.mutable.tileentity.BannerData;
+import org.spongepowered.api.data.meta.PatternLayer;
+import org.spongepowered.api.data.type.DyeColor;
+import org.spongepowered.api.service.persistence.InvalidDataException;
+import org.spongepowered.common.data.manipulator.mutable.tileentity.SpongeBannerData;
+import org.spongepowered.common.data.util.DataUtil;
+
+import com.google.common.base.Optional;
+
+public class BannerDataBuilder implements DataManipulatorBuilder<BannerData, ImmutableBannerData> {
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Optional<BannerData> build(DataView container) throws InvalidDataException {
+        DataUtil.checkDataExists(container, Keys.BANNER_BASE_COLOR.getQuery());
+        DataUtil.checkDataExists(container, Keys.BANNER_PATTERNS.getQuery());
+        final DyeColor baseColor = (DyeColor) container.get(Keys.BANNER_BASE_COLOR.getQuery()).get();
+        final List<PatternLayer> patterns = (List<PatternLayer>) container.getList(Keys.BANNER_PATTERNS.getQuery()).get();
+        return Optional.<BannerData>of(new SpongeBannerData(baseColor, patterns));
+    }
+
+    @Override
+    public BannerData create() {
+        return new SpongeBannerData();
+    }
+
+    @Override
+    public Optional<BannerData> createFrom(DataHolder dataHolder) {
+        // TODO 
+        if (dataHolder instanceof TileEntityBanner) {
+            final Color baseColor;
+            final List<TileEntityBanner.EnumBannerPattern> mojangBannerPattern;
+            final List<EnumDyeColor> mojangEnumDyeColor;
+        }
+        return null;
```

Comment:
You need to return absent. Also, you need to consider itemstacks having the banner data (which is existent in 1.8).